### PR TITLE
Add UPI (HDFC) Circuits

### PR DESCRIPTION
**Pull request by 0xSachinK**
### Description

Notes
* Nullifier: hashed payment ID for smart contract processors. Hashed payment ID is strictly better than email hash as nullifier
* For client, we have created a `replaceMessageIdWithXGoogleOriginalMessageId` function in the preprocess_input.ts script that the frontend can use (included in npm package) @asoong 

####  Regexes
- [x] HDFC Account Number
- [x] HDFC Amount
- [x] HDFC Date
- [x] HDFC Payee ID
- [x] HDFC UPI Subject
- [x] To (email) Regex

#### Circuits
- [x] HDFC Registration
- [x] HDFC Send 

### Checklist

For any circuit changes, complete the following steps to ensure faster integration with UI and Smart Contracts:

- [x] Ran `yarn:compile:XXX` where `XXX` is the name of the circuit
- [x] Commit the build files (build/venmo_*/venmo_*_js/*.js and build/venmo_*/venmo_*_js/*.wasm) generated by the above command
- [x] Ran `yarn test` to check all tests are passing
- [x] Ran `yarn:genkey:chunked:unsafe:XXX` AND `yarn:genkey:non-chunked:unsafe:XXX` to generate new proving keys for each circuit that was updated
- [x] Updated `UPLOAD_FOLDER` in `circuit.env` to a new path in S3. E.g. `UPLOAD_FOLDER="v2/v0.0.5"`
- [x] Ran `yarn uploadkeys:XXX` to upload to S3. Ensure that BOTH chunk and nonchunk keys are uploaded
- [x] Ran `yarn:genverifier:XXX` to generate solidity verifier
- [x] Copy the generated verifier and paste it in the `contracts` folder
- [x] Ran `yarn:gen-input:XXX` to generate inputs for witness generation
- [x] Ran `yarn:genwitness:XXX` to generate witnesses for proof generation
- [x] Ran `yarn:genproof:XXX` to generate proof for test solidity calldata
- [x] Ran `yarn:gencalldata:XXX` to generate solidity calldata
- [x] Paste test calldata below
- [x] Ensured files are public through ACL in `zk-p2p` S3 bucket
- [x] Updated `package.json` in circuits to correct version and run `npm publish`
- [x] Commit the updated `package.json`

#### HDFC Registration

["0x1ac7f73df1c3e588f3f2e737db7687c48c0714feb11fd75d7338213b4bfe864e", "0x1df991c9b756f71237b83bbdfda00b41b66a05faa0e13f9cac9c2f57cbc7cc70"],[["0x1a1b2e29898f8432944f3c01ac132b35cc0d69d40beabea61174681491cfee03", "0x25e155983d41ab5d8d86b082dc35699e46f805cfafdf6813e42a6ec78b9bc446"],["0x2dedba03e1eec70caad450667dfaadd3d73d465c4c476245d445cae6ce3eee42", "0x04c348e01dc6566775621d90807b30937604bd5701befca9d63f70610d0bd1f5"]],["0x21d1429cdf748fb0ef18f11d928240934d7795c83b887e5740bfceff20a7f7e4", "0x1a4e18301da5746741f127f50aa2604bdcdb666b5b7655a4234f6bbc4515fe47"],["0x06b0ad846d386f60e777f1d11b82922c6bb694216eed9c23535796ac404a7dfa","0x0000000000000000000000000000000000000000000000000040737472656c61","0x000000000000000000000000000000000000000000000000006e616263666468","0x00000000000000000000000000000000000000000000000000000074656e2e6b","0x2282c0b9cd1bedb8f14f72c2c434886a10b0c539ad1a5d62041c4bfa3ef5c7c7"]
#### HDFC Send

["0x01334bb73919e2905a4610c0367b0100a9fd34688bb39203b0fcfc2debdc9806", "0x27f52a89e2e05e0f55d2ac70df0b1170f252d3d537020f4e19117361690bffb7"],[["0x1e769479e9ac6ea605c6191175b5bacaf11050916e707a8ba5c5c1390789b9cb", "0x00baf45aa456d7f0503204b910bfa50136afd383fbe7d3d398b860a3c18338c2"],["0x2f35b4c36bb9025737526fcfcf73288d0207871cb105ad7d1c06215a80a3842b", "0x2873d615f840baa48f643a790ee2c91d28f866b655e08da8c818184b4563176c"]],["0x0900598759487cbd9bacc5a4a9d34090bf4439cf8200ecb56927538cda6ea361", "0x2a6ca6465e769cbc5ca20d2838e14543d21ec677a411d328e9d9606d5b3a535c"],["0x06b0ad846d386f60e777f1d11b82922c6bb694216eed9c23535796ac404a7dfa","0x0000000000000000000000000000000000000000000000000040737472656c61","0x000000000000000000000000000000000000000000000000006e616263666468","0x00000000000000000000000000000000000000000000000000000074656e2e6b","0x0000000000000000000000000000000000000000000000000000000030302e32","0x0000000000000000000000000000000000000000000000000000000000000000","0x000000000000000000000000000000000000000000000000003132202c657554","0x00000000000000000000000000000000000000000000000000303220766f4e20","0x00000000000000000000000000000000000000000000000000353a3131203332","0x00000000000000000000000000000000000000000000000000302b2038303a38","0x0000000000000000000000000000000000000000000000000000000000303335","0x2282c0b9cd1bedb8f14f72c2c434886a10b0c539ad1a5d62041c4bfa3ef5c7c7","0x2e77b67e4f7d868c763d4539ca8483672e522d3aafb015f27aef9090f6790a18","0x0bd3bb4a2060e5b6d300226add771815983afad1c596b4cf191639ccd74874a9","0x0000000000000000000000000000000000000000000000000000000000003039"]
#### New Package Version

0.1.1-rc6

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE/circuit_template.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ For any circuit changes, complete the following steps to ensure faster integrati
 - [ ] Ran `yarn:compile:XXX` where `XXX` is the name of the circuit
 - [ ] Commit the build files (build/venmo_*/venmo_*_js/*.js and build/venmo_*/venmo_*_js/*.wasm) generated by the above command
 - [ ] Ran `yarn test` to check all tests are passing
+- [ ] Update the `circuit.env` with the new `PTAU` power
```
Comment (richardliang) [inline .github/PULL_REQUEST_TEMPLATE/circuit_template.md:13]: What's this?

File: circuits-circom/hdfc_registration.circom [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+include "./regexes/from_regex.circom";
+include "./regexes/hdfc/hdfc_upi_subject.circom";
+
+
+template HdfcRegistrationEmail(max_header_bytes, max_body_bytes, n, k, pack_size) {
+    assert(n * k > 1024); // constraints for 1024 bit RSA
+
+    // Rounded to the nearest multiple of pack_size for extra room in case of change of constants
+    var max_email_from_len = ceil(21, pack_size); // RFC 2821: requires length to be 254, but we can limit to 21 (alerts@hdfcbank.net)
+    var max_email_to_len = ceil(35, pack_size); // TODO: Change this.
```
Comment (richardliang) [inline circuits-circom/hdfc_registration.circom:19]: for this, think we need to use RFC2821? So support max length of 254

File: circuits-circom/hdfc_registration.circom [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    var max_account_number_len = ceil(7, pack_size);    // TODO: Change this. Max is only 4 digits but set to 7 for now.
```
Comment (richardliang) [inline circuits-circom/hdfc_registration.circom:20]: Update todo?

File: circuits-circom/hdfc_registration.circom [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    signal output onramper_id <== HdfcOnramperId(
+        max_email_to_len, 
+        max_header_bytes, 
+        max_account_number_len, 
+        max_body_bytes, 
+        pack_size
+    )(in_padded, email_to_idx, in_body_padded, hdfc_acc_num_idx);
+
+    // TOTAL CONSTRAINTS: (WITH STUB): 936240
+    // WITHOUT STUB: TODO
```
Comment (richardliang) [inline circuits-circom/hdfc_registration.circom:77]: Let's recompile with the stub. Any reason not to?

File: circuits-circom/hdfc_send.circom [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+include "./regexes/from_regex.circom";
+include "./regexes/hdfc/hdfc_amount.circom";
+include "./regexes/hdfc/hdfc_date.circom";
+include "./regexes/hdfc/hdfc_upi_subject.circom";
+
+template HdfcSendEmail(max_header_bytes, max_body_bytes, n, k, pack_size) {
+    assert(n * k > 1024); // constraints for 1024 bit RSA
+
+    // Rounded to the nearest multiple of pack_size for extra room in case of change of constants
+    var max_email_amount_len = 8; // Allowing max 4 fig amount+ one comma  + one decimal point + 2 decimal places. e.g. 2,500.00
```
Comment (richardliang) [inline circuits-circom/hdfc_send.circom:21]: What is the max limit for UPI? Currently this supports 9,999.99 INR as the max which isn't that high

File: circuits-circom/hdfc_send.circom [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    var max_email_from_len = ceil(21, pack_size); // RFC 2821: requires length to be 254, but we can limit to 21 (alerts@hdfcbank.net)
+    var max_email_date_len = 31; // Sat, 14 Oct 2023 22:09:12 +0530
+    // TODO: CHANGE THIS TO 50.
```
Comment (richardliang) [inline circuits-circom/hdfc_send.circom:24]: Lets change this

File: circuits-circom/hdfc_send.circom [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    var max_email_to_len = ceil(35, pack_size); // TODO: Change this.
+    var max_payee_len = ceil(42, pack_size);    // Max 50 characters in UPI ID  (TODO: CHANGE THIS TO 50)
```
Comment (richardliang) [inline circuits-circom/hdfc_send.circom:26]: Update to 50

File: circuits-circom/regexes/hdfc/hdfc_accnum.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{
+  "parts": [
+    {
+      "is_public": false,
+      "regex_def": "\\*\\*"
```
Comment (richardliang) [inline circuits-circom/regexes/hdfc/hdfc_accnum.json:5]: Maybe add a line break \r\n to make it a bit safer

File: circuits-circom/regexes/hdfc/hdfc_payee_id.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
+{
+  "parts": [
+    {
+      "is_public": false,
+      "regex_def": "to VPA "
```
Comment (richardliang) [inline circuits-circom/regexes/hdfc/hdfc_payee_id.json:5]: Let's actually consolidate payee_id and acc_num into 1 regex string with 2 reveals

File: circuits-circom/test/hdfc_registration.spec.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+        const combinedArray = packed_to_email_array.concat(packed_account_number_array)
+        const expected_hash = poseidon(combinedArray);
+        const expected_hash_contract = await poseidonContract["poseidon(uint256[6])"](combinedArray);
+
+        assert.equal(JSON.stringify(poseidon.F.e(hashed_registration_id)), JSON.stringify(expected_hash), true);
+        assert.equal(JSON.stringify(poseidon.F.e(hashed_registration_id)), JSON.stringify(poseidon.F.e(expected_hash_contract.toString())), true);
+    });
+
+    // Could not find another email from alerts@hdfcbank.net that fits inside the 3200 limit set by us for the UPI email
```
Comment (richardliang) [inline circuits-circom/test/hdfc_registration.spec.ts:172]: Is this still relevant? Do we need to increase the 3200 limit?

File: circuits-circom/hdfc_registration.circom (hunk 2 above)
Comment (0xSachinK) [reply circuits-circom/hdfc_registration.circom:19]: Do we want to? Is there a safe max email length that we can assume?

File: circuits-circom/hdfc_registration.circom (hunk 4 above)
Comment (0xSachinK) [reply circuits-circom/hdfc_registration.circom:77]: Not necessary while we are still in dev and iterating on circuits.

File: circuits-circom/regexes/hdfc/hdfc_accnum.json (hunk 8 above)
Comment (0xSachinK) [reply circuits-circom/regexes/hdfc/hdfc_accnum.json:5]: There wouldn't always be a line break.

File: circuits-circom/regexes/hdfc/hdfc_payee_id.json (hunk 9 above)
Comment (0xSachinK) [reply circuits-circom/regexes/hdfc/hdfc_payee_id.json:5]: I was thinking about it. Although why is the current one not good? This is a tradeoff between cleaner and abstracted code with more constraints vs not cleaner code with less constraints.

File: circuits-circom/hdfc_registration.circom (hunk 2 above)
Comment (0xSachinK) [reply circuits-circom/hdfc_registration.circom:19]: Set it to 56 which is a safe length. Source: https://atdata.com/long-email-addresses/

**Comment by pratham-bhatnagar** [comment]
Hey @0xSachinK,
Im very interested in contributing ZKp2p. I love to see HDFC and UPI around. How can I contribute in the project as well?

**Pull request merged**